File: 1_money_change_problem/change_dp.py
# Uses python3
import sys


def get_change(money):
    coins = [1,3,4]
    minNumCoins = list()
    minNumCoins.append(0)
    for m in range(1, money+1):
        minNumCoins.append(sys.maxsize)
        for i in coins:
            if m>=i:
                numCoins = minNumCoins[m-i] + 1
                if numCoins < minNumCoins[m]:
                    minNumCoins[m] = numCoins
                
    return minNumCoins[m]

if __name__ == '__main__':
    m = int(sys.stdin.read())
    print(get_change(m))

File: 1_money_change_problem/change_dp.cpp
#include <iostream>
#include <climits>
#include <vector>

int get_change(int money) {
  std::vector<int> minNumCoins ;
  minNumCoins.push_back(0);
  int coins[3] = {1,3,4};
  int numCoins;
  for(int m = 1; m <= money ; m++)
    {
      minNumCoins.push_back(INT_MAX);
      for(int i = 0; i < 3; i++)
	{
	  if(m >= coins[i])
	    {
	      numCoins = minNumCoins[m - coins[i]] + 1;
	      if(numCoins < minNumCoins[m])
		{
		  minNumCoins[m] = numCoins;
		}
	      
	    }
	}
    }
  return minNumCoins[money];
}

int main() {
  int m;
  std::cin >> m;
  std::cout << get_change(m) << '\n';
}
